src/field/vecops.cxx

```diff
@@ -212,70 +262,89 @@ const Vector2D Curl(const Vector2D &v, CELL_LOC outloc) {
+    outloc = v.getLocation();
+  }
+
+  // We can't support VSHIFT here as, e.g. DDY can't produce an output at CELL_XLOW
+  // unless the input field is at CELL_XLOW, but then that field will also be needed
+  // at CELL_YLOW, for example for another component.
+  ASSERT1(outloc != CELL_VSHIFT);
+
   Mesh *localmesh = v.x.getMesh();
-  Coordinates *metric = localmesh->coordinates(outloc);
+  auto metric = localmesh->coordinates(outloc);
 
   // Get covariant components of v
   Vector2D vco = v;
   vco.toCovariant();
 
   // get components (curl(v))^j
   Vector2D result(localmesh);
-  result.x = (DDY(vco.z, outloc) - DDZ(vco.y, outloc))/metric->J;
-  result.y = (DDZ(vco.x, outloc) - DDX(vco.z, outloc))/metric->J;
-  result.z = (DDX(vco.y, outloc) - DDY(vco.x, outloc))/metric->J;
+  result.x = (DDY(vco.z, outloc) - DDZ(vco.y, outloc)) / metric->J;
+  result.y = (DDZ(vco.x, outloc) - DDX(vco.z, outloc)) / metric->J;
+  result.z = (DDX(vco.y, outloc) - DDY(vco.x, outloc)) / metric->J;
 
   /// Coordinate torsion
-  result.z -= metric->ShiftTorsion*vco.z / metric->J;
+  result.z -= metric->ShiftTorsion * vco.z / metric->J;
+
+  result.setLocation(outloc);
 
   result.covariant = false; // result is contravariant
 
   return result;
 }
 
-const Vector3D Curl(const Vector3D &v, 
-                    CELL_LOC outloc_x, CELL_LOC outloc_y, CELL_LOC outloc_z) {
+const Vector3D Curl(const Vector3D &v, CELL_LOC outloc_x, CELL_LOC outloc_y,
+                    CELL_LOC outloc_z) {
+  TRACE("Curl( Vector3D )");
+  ASSERT1(outloc_x == outloc_y && outloc_x == outloc_z);
+  return Curl(v, outloc_x);
+}
 
+const Vector3D Curl(const Vector3D &v, CELL_LOC outloc) {
   TRACE("Curl( Vector3D )");
 
-  Mesh *localmesh = v.x.getMesh();
+  if (outloc == CELL_DEFAULT) {
+    outloc = v.getLocation();
+  };
 
-  Coordinates* metric_z = localmesh->coordinates(outloc_z);
+  // We can't support VSHIFT here as, e.g. DDY can't produce an output at CELL_XLOW
+  // unless the input field is at CELL_XLOW, but then that field will also be needed
+  // at CELL_YLOW, for example for another component.
+  ASSERT1(outloc != CELL_VSHIFT);
+
+  Mesh *localmesh = v.x.getMesh();
+  auto metric = v.x.getCoordinates(outloc);
 
   // Get covariant components of v
   Vector3D vco = v;
   vco.toCovariant();
 
   // get components (curl(v))^j
   Vector3D result(localmesh);
-  result.x = (DDY(vco.z, outloc_x) - DDZ(vco.y, outloc_x))/localmesh->coordinates(outloc_x)->J;
-  result.y = (DDZ(vco.x, outloc_y) - DDX(vco.z, outloc_y))/localmesh->coordinates(outloc_y)->J;
-  result.z = (DDX(vco.y, outloc_z) - DDY(vco.x, outloc_z))/metric_z->J;
+  result.x = (DDY(vco.z, outloc) - DDZ(vco.y, outloc)) / metric->J;
+  result.y = (DDZ(vco.x, outloc) - DDX(vco.z, outloc)) / metric->J;
+  result.z = (DDX(vco.y, outloc) - DDY(vco.x, outloc)) / metric->J;
 
   // Coordinate torsion
-  result.z -= metric_z->ShiftTorsion*vco.z / metric_z->J;
+  result.z -= metric->ShiftTorsion * vco.z / metric->J;
```

Comment:
I think we have to have `vco.z.getLocation()==outloc` here, because otherwise one of the derivatives above would throw, but it's not asserted in this function. I think we should either assert than `v.getLocation()==outloc` or put `interp_to(vco.z, outloc)` here.

Reply:
I'm not sure I follow -- which combinations of input and output locations do we support here? Is it just the usual one staggered and one not? If so, then one of the derivatives here should catch problems. If they have to be at the same location, we should not support `outloc` at all.